keps/sig-node/2570-memory-qos/README.md

```diff
@@ -104,6 +107,145 @@ Brief map as follows:
+```
+memory.high=floor[(requests.memory + memory throttling factor * (limits.memory or node allocatable memory - requests.memory))/pageSize] * pageSize, where default value of memory throttling factor is set to 0.9
+```
+Note: If a container had no memory limit specified, we substitute `limits.memory` for `node allocatable memory` and apply the throttling factor of .9 to that value.
+
+
+The table below runs over the examples with different values requests.memory:
+
+| limits.memory (1000) | memory throttling factor (0.9)|
+| ---------------------- | ----------------------------- |
+| request 0    | 900  |
+| request 100  | 910  |
+| request 200  | 920  |
+| request 300  | 930  |
+| request 400  | 940  |
+| request 500  | 950  |
+| request 600  | 960  |
+| request 700  | 970  |
+| request 800  | 960  |
+| request 900  | 980  |
+| request 1000 | 1000 |
+
+
+Node reserved resources(kube-reserved/system-reserved) are either considered. It is tied to `--enforce-node-allocatable` and `memory.min` will be set properly.
+
+Brief map as follows:
+| type | memory.min | memory.high |
+| -------- | -------- | -------- | 
+| container | requests.memory | floor[(requests.memory + memory throttling factor * (limits.memory or node allocatable memory - requests.memory))/pageSize] * pageSize |
+| pod | sum(requests.memory) | n |
+| node | n/a | pods, kube-reserved, system-reserved | n |
+
+###### Reasons for changing the formula of memory.high calculation in Beta
+
+The formula for memory.high has changed in Beta as the Alpha implementation has following problems:
+1. It fails to throttle when requested memory is closer to memory limits (or node allocatable): memory.high < requests.memory. e.g. If `requests.memory = 85, limits.memory=100`, and we have a throttling factor of 0.8, then as per the Alpha implementation memory.high =  memory throttling factor * limits.memory i.e. memory.high = 80. In this case the level at which throttling is supposed to occur i.e. memory.high is less than requests.memory. Hence there won't be any throttling as the Alpha implementation doesn't allow memory.high to be less than requested memory. 
+
+2. It could result in early throttling even when a sizeable chunk of memory is unused. 
+  
+    For example, 
+    * `requests.memory` = 800Mi
+
+      `memory throttling factor` = 0.8
+
+      `limits.memory` = 1000Mi
+
+      As per Alpha implementation,
+
+      `memory.high` = memory throttling factor * limits.memory  = 0.8 * 1000Mi = 800Mi
+
+      Throttling occurs at 800Mi, while a sizable chunk of 200Mi is still unused.
+
+    * `requests.memory` = 500Mi
+
+      `memory throttling factor` = 0.6
+
+      `limits.memory` = 1000Mi
+      
+      As per Alpha implementation,
+
+      `memory.high` = memory throttling factor * limits.memory = 0.6 * 1000Mi = 600Mi
+      
+      Throttling occurs at 600Mi which is just a 100Mi over the requested memory. 400Mi of memory is still unused as early throttling happens.
```

Comment:
thanks for adding this example, it helps illustrate the changes

Reply:
I agree that we should punish users if the memory usage is above the request. But with `memory.high`, the throttling is too aggressive.

> Another solution is like koordinator 
> https://github.com/koordinator-sh/koordinator/blob/e51758ce22b9590ab71219f6bd75e8847279b049/apis/slo/v1alpha1/nodeslo_types.go#L35-L39 or https://help.aliyun.com/document_detail/413260.html.
> They introduced a new cgroup param that is not in official cgroup v2: memory.wmark_high
> 
> When the memory is over the memory.wmark_hgih, it will trigger reclaiming memory.
> And they have a wmarkRatio that by default is 95. 
> - They disable throttling by default
> -  - The memory.wmark_high is by default 95% of memory.limit
> -  When user enable throttling,
> -  - The memory.wmark_high is by default 95% of memory.high
![image](https://user-images.githubusercontent.com/2010320/216743599-a3d0986d-9d5b-497c-9e1a-79d1e5517488.png)

if the `memory.high` is like `memory.wmark_high` mentioned above, I would prefer to set it to a lower value. However, throttling is a behavior change, so I prefer to set it higher at first, especially that we may one day make it default behavior.

Reply:
ACK - let's just be clear in how we describe this.  If the throttling behavior is too aggressive, it may not be what we need :)

Reply:
Thanks @thockin, for pointing this out, I agree, we should rephrase this a bit and clarify the overall intent is that containers should ideally not exceed their requests -- otherwise it points that the apps requests are too low in the first place.

> Every app that uses memory > request puts the whole node at risk of overcommit. I think it's a feature that the OS punishes containers that are over their request with aggressive reclaim. Ideally that would ramp up as you get farther and farther from the request.

> Apps which go over-request and stay there are misconfigured.

This is definitely the case, but the concern is that many applications are misconfigured today because setting accurate requests and limits is really challenging. Even if you do set them right for most cases, if there is a sudden load increase they will be incorrect. 

I think the ideal long term vision is a world in where users don't need to set request and limits at all and instead something like VPA++ could do it automatically based on usage and cpu/memory pressure signals. In fact, the kernel already provides us such as signals via cgroupv2 [PSI metrics](https://facebookmicrosites.github.io/psi/docs/overview). So for example, if memory pressure is low -> VPA needs to decrease the resources requested; if the memory pressure is high, maybe bump the limit. @vinaykul work on in resize KEP could allow us to update these settings in place :) 

But coming back to today's situation, we need a bit careful to avoid inducing throttling too early on user's applications, so it's why we stayed a bit on the conservative side and have a high throttling factor so memory.high is closer to memory limit as opposed to memory request. Users can always adjust the throttling factor based on their risk tolerance.

Reply:
> ACK - let's just be clear in how we describe this. If the throttling behavior is too aggressive, it may not be what we need :)

@thockin Yes, I prefer to not throttle by default. And the current alpha feature uses a 0.8 factor and that is why I want to change it to 0.9 as this is the default behavior. Changing the factor to bigger will prevent the throttling too early for users.
- But as Bobbay pointed out, many applications are misconfigured today because setting accurate requests and limits is really challenging. 
- Plus, I also asked for a way for pods to opt-in to this feature. However, the current answer would be that a performance-sensitive or high-priority pod should use guaranteed memory configuration.

I am not sure if the watermark ratio that I mentioned is in the progress of kernel teams.
I would appreciate it if we set the default throttling factor to 1 which means no throttling by default, and set a  watermark ratio to 0.8 or 0.9 to trigger the asynchronous memory reclaim once the watermark ratio is supported on the kernel side.

Reply:
I said:

What's the point of throttling at limit, when we know OOM will handle it?

I'd there's a way to apply pressure (more aggressive reclaim) between
request and limit, that sounds like a win

@pacoxu said

> What's the point of throttling at a limit, when we know OOM will handle it? I'd there's a way to apply pressure (more aggressive reclaim) between request and limit, that sounds like a win

I agree that for most cases, throttling will help pods not to be OOMed. 

There is a special case for some Java programs that will always be in a high memory but it(JVM) has its way to do garbage collections. 
- With a sudden increase in the amount of access/application QPS, the memory will remain high for a period, and HPA may be triggered at the same time to start more pods. 
- BTW, a high memory in most time indicts the service is under large pressure and throttling may make it worse in the response time. OOM should help us found the memory leak part and in some cases, they need to add some rate limit logic on the gateway or codes.

If a throttle takes effect in such a scenario, it would be not expected. [of course, we can suggest such users use a higher request to avoid the throttling, users may struggle].


OOM is service termination, and throttling is service degradation.
- limit is a hard limit.
- throttling is a soft limit.

If we set this soft limit for a pod, I would suggest adding this soft limit info to the pod status so that users can know when this pod will be throttled. Different nodes use different throttling factors as we designed here.

Reply:
You said earlier "I would appreciate it if we set the default throttling factor to 1 which means no throttling by default" I just wanted to point out that the point of that setting doesn't DO anything - or does it?

> a high memory in most time indicts the service is under large pressure and throttling may make it worse in the response time

Yes, which is why a simple threshold is imperfect.  If you go above the line for a short period of time, it should be fine.  If you stay there, it's a problem.

Let's remember - the point of cgroups is to protect containers from each other.  Allowing users to set limit > request (for non-compressible resources) totally violates that goal.  It should not be seen as "normal" to set limit significantly higher than request, and pods that do that SHOULD be squeezed.

The cgroup docs say "the cgroup’s processes are throttled and put under heavy reclaim pressure".  It doesn't say HOW heavy.

The singular "high" setting which leads to "heavy reclaim pressure" may simply be too coarse to be very useful.  It feels to me that we want a "gentle reclaim" at `request`, with increasing aggressiveness the farther you get from it, and ultimately OOM at `limit`.

>  I would suggest adding this soft limit info to the pod status

I am all for status, as long as it's not overly tied to implementation details - there are other runtimes out there and who knows, when cgroup v3 lands (or the v2 details change) our API can end up broken.

> I also asked for a way for pods to opt-in to this feature

I'll repeat my thesis: the point of cgroups is to protect containers from each other.  Allowing them to opt-in to things like this defats the purpose.

Reply:
Do we have an experiment and graphs that show the amount of reclaim done on a cgroup as it moves from below-request to above-request to throttling ?  It would be very informational.

Reply:
Re: special case JVM applications, IMO we should recommend using guaranteed pods as per the suggested implementation. Guaranteed pods won't be throttled and can use memory up to the set limit.

Reply:
> What's the point of throttling at limit, when we know OOM will handle it?
> If there's a way to apply pressure (more aggressive reclaim) between
request and limit, that sounds like a win

I think there is two pieces here:

1.  Help prevent an OOM from possibly happening in the first place by starting heavy reclaim a bit before the limit is reached. The hope with doing that is it provides an opportunity to decrease the memory usage and hopefully prevent the OOM from happening. 

A concrete use case is https://github.com/kubernetes/kubernetes/issues/43916. The summary of that longstanding issue is that some apps read many files into memory, filling the page cache. Eventually, this may result in cgroup OOM even though the memory could be reclaimed. Some users have written workarounds like agents that proactively force reclaim https://github.com/linchpiner/cgroup-memory-manager. The goal is that with this feature, this will not be necessary, as  `memory.high` (a bit before the limit) would be reached, and the memory would be reclaimed automatically. 

2. The second point regarding more aggressively applying pressure between request and limit is something we want as well, since it discourages bad behavior of going over request and also provides better estimates around what the current memory usage is for things like VPA. The kernel very recently (6 months) got a knob for this https://github.com/torvalds/linux/commit/94968384dde15d48263bfc59d280cd71b1259d8c, in so called `memory.reclaim` cgroupv2 API. This allows userspace to request reclaim of X amount in a given cgroup to be reclaimed. Using that functionally, we could accomplish gradually increasing reclaim as pod's memory usage will go over requests. This is future work though and will need its own separate KEP and we should split it from first use case.